src/per/uart.cpp

```diff
@@ -34,135 +67,371 @@ struct uart_handle
+pin_alt usart1_pins_tx[] = {{{DSY_GPIOB, 6}, GPIO_AF7_USART1},
+                            {{DSY_GPIOB, 14}, GPIO_AF4_USART1},
+                            pins_none};
+pin_alt usart1_pins_rx[] = {{{DSY_GPIOB, 7}, GPIO_AF7_USART1},
+                            {{DSY_GPIOB, 15}, GPIO_AF4_USART1},
+                            pins_none};
+
+pin_alt usart2_pins_tx[]
+    = {{{DSY_GPIOA, 2}, GPIO_AF7_USART2}, pins_none, pins_none};
+pin_alt usart2_pins_rx[]
+    = {{{DSY_GPIOA, 3}, GPIO_AF7_USART2}, pins_none, pins_none};
+
+pin_alt usart3_pins_tx[]
+    = {{{DSY_GPIOC, 10}, GPIO_AF7_USART3}, pins_none, pins_none};
+pin_alt usart3_pins_rx[]
+    = {{{DSY_GPIOC, 11}, GPIO_AF7_USART3}, pins_none, pins_none};
+
+pin_alt uart4_pins_tx[] = {{{DSY_GPIOB, 9}, GPIO_AF8_UART4},
+                           {{DSY_GPIOC, 10}, GPIO_AF8_UART4},
+                           pins_none};
+pin_alt uart4_pins_rx[] = {{{DSY_GPIOB, 8}, GPIO_AF8_UART4},
+                           {{DSY_GPIOC, 11}, GPIO_AF8_UART4},
+                           pins_none};
+
+pin_alt uart5_pins_tx[] = {{{DSY_GPIOC, 12}, GPIO_AF8_UART5},
+                           {{DSY_GPIOB, 6}, GPIO_AF14_UART5},
+                           pins_none};
+pin_alt uart5_pins_rx[] = {{{DSY_GPIOB, 12}, GPIO_AF14_UART5},
+                           {{DSY_GPIOD, 2}, GPIO_AF8_UART5},
+                           {{DSY_GPIOB, 5}, GPIO_AF14_UART5}};
+
+pin_alt usart6_pins_tx[] = {pins_none, pins_none, pins_none};
+pin_alt usart6_pins_rx[]
+    = {{{DSY_GPIOG, 9}, GPIO_AF7_USART6}, pins_none, pins_none};
+
+pin_alt uart7_pins_tx[]
+    = {{{DSY_GPIOB, 4}, GPIO_AF11_UART7}, pins_none, pins_none};
+pin_alt uart7_pins_rx[] = {pins_none, pins_none, pins_none};
+
+pin_alt uart8_pins_tx[] = {pins_none, pins_none, pins_none};
+pin_alt uart8_pins_rx[] = {pins_none, pins_none, pins_none};
+
+pin_alt lpuart1_pins_tx[]
+    = {{{DSY_GPIOC, 6}, GPIO_AF8_LPUART}, pins_none, pins_none};
+pin_alt lpuart1_pins_rx[]
+    = {{{DSY_GPIOC, 7}, GPIO_AF8_LPUART}, pins_none, pins_none};
+
+//an array to hold everything
+pin_alt* pins_periphs[] = {usart1_pins_tx,
+                           usart1_pins_rx,
+                           usart2_pins_tx,
+                           usart2_pins_rx,
+                           usart3_pins_tx,
+                           usart3_pins_rx,
+                           uart4_pins_tx,
+                           uart4_pins_rx,
+                           uart5_pins_tx,
+                           uart5_pins_rx,
+                           usart6_pins_tx,
+                           usart6_pins_rx,
+                           uart7_pins_tx,
+                           uart7_pins_rx,
+                           uart8_pins_tx,
+                           uart8_pins_rx,
+                           lpuart1_pins_tx,
+                           lpuart1_pins_rx};
+
+bool gpiopin_equal(dsy_gpio_pin a, dsy_gpio_pin b)
+{
+    return a.port == b.port && a.pin == b.pin;
+}
```

Comment:
The `dsy_gpio_cmp` function in `daisy_core.h` already exists. You should be able to use that instead.